pandas/core/arrays/categorical.py

```diff
@@ -659,6 +661,22 @@ def _codes_for_groupby(self, sort):
             categories in the original order.
         """
 
+        # we only care about observed values
+        if observed:
+            unique_codes = unique1d(self.codes)
+            cat = self.copy()
+
+            take_codes = unique_codes[unique_codes != -1]
+            if self.ordered:
+                take_codes = np.sort(take_codes)
+
+            # we recode according to the uniques
+            cat._categories = self.categories.take(take_codes)
```

Comment:
To explain my confusion: what this returns is "strange", as it does returns a modified categorical that does not match the data any longer. Eg for

```
cat = pd.Categorical(['a', 'c', 'a'], categories=['a', 'b', 'c'])
```

this returns a categorical with values ['a', 'b', 'a'], because you recode the codes (based on observed categories) without changing the categories. 

This makes that the `Grouping._group_index` is "incorrect", but apparently this is fine for the actual groupby implementation because the groupby results are correct (but, it makes the above function somewhat confusing). But with playing around with it, I found a place where it is actually propagated:

```
In [52]: cat = pd.Categorical(['a', 'c', 'a'], categories=['a', 'b', 'c'])

In [53]: df = pd.DataFrame({'cat': cat, 'vals': [1, 2, 3]})

In [54]: g = df.groupby('cat', observed=True)

In [55]: g.sum()   <--- correct
Out[55]: 
     vals
cat      
a       4
c       2

In [56]: g.groups    <--- incorrect
Out[56]: 
{'a': Int64Index([0, 2], dtype='int64'),
 'b': Int64Index([1], dtype='int64'),
 'c': Int64Index([], dtype='int64')}
```

(same for `g.indices`)